File: 347-top-k-frequent-elements/top-k-frequent-elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function(nums, k) {
    // add to freq map
    const freqMap = {};
    for(let num of nums) {
        if(freqMap[num]) {
            freqMap[num]++;
        } else {
            freqMap[num] = 1;
        }
    }

    const returnVal = [];

    // Sort the map desc based on freq
    const entries = Object.entries(freqMap).sort((a,b) => b[1] - a[1]);
    for(let i=0; i<k; i++) {
        returnVal.push(Number(entries[i][0]));
    }

    return returnVal;
};

File: 347-top-k-frequent-elements/top-k-frequent-elements.cpp
class Solution {
    typedef pair<int, int> P;

public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> freq;
        for(int num: nums) { // O(N)
            freq[num]++;
        }

        priority_queue<P, vector<P>, greater<P>> minHeap;
        for(auto& entry: freq) { // O(k)
            minHeap.push({entry.second, entry.first}); // O(log k)
            if(minHeap.size() > k) minHeap.pop();
        }

        vector<int> ans;
        while(!minHeap.empty()) { // O(k)
            int top = minHeap.top().second;
            ans.push_back(top); 
            minHeap.pop(); // O(log k)
        }

        return ans;
    }
};